extensions-core/s3-extensions/src/test/java/org/apache/druid/storage/s3/output/RetriableS3OutputStreamTest.java

```diff
@@ -0,0 +1,382 @@
+    new RetriableS3OutputStream(config, s3, path).close();
+  }
+
+  @Test
+  public void testTooSmallChunkSizeMaxResultsSizeIsNotRetionalToMaxPartNum() throws IOException
+  {
+    maxResultsSize = 274877906944L;
+    chunkSize = 27487790;
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage(
+        "chunkSize[27487790] is too small for maxResultsSize[274877906944]. chunkSize should be at least [27487791]"
+    );
+    new RetriableS3OutputStream(config, s3, path).close();
+  }
+
+  @Test
+  public void testTooLargeChunkSize() throws IOException
+  {
+    maxResultsSize = 1024L * 1024 * 1024 * 1024;
+    chunkSize = RetriableS3OutputStream.S3_MULTIPART_UPLOAD_MAX_PART_SIZE + 1;
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage(
+        "chunkSize[5368709121] should be smaller than [5368709120]"
+    );
+    new RetriableS3OutputStream(config, s3, path).close();
+  }
+
+  @Test
+  public void testWriteAndHappy() throws IOException
+  {
+    maxResultsSize = 1000;
+    chunkSize = 10;
+    ByteBuffer bb = ByteBuffer.allocate(Integer.BYTES);
+    try (RetriableS3OutputStream out = new RetriableS3OutputStream(
+        config,
+        s3,
+        path,
+        false
+    )) {
+      for (int i = 0; i < 25; i++) {
+        bb.clear();
+        bb.putInt(i);
+        out.write(bb.array());
+      }
+    }
+    // each chunk is 10 bytes, so there should be 10 chunks.
+    Assert.assertEquals(10, s3.partRequests.size());
+    s3.assertCompleted(chunkSize, Integer.BYTES * 25);
+  }
+
+  @Test
+  public void testWriteSizeLargerThanConfiguredMaxChunkSizeShouldSucceed() throws IOException
+  {
+    maxResultsSize = 1000;
+    chunkSize = 10;
+    ByteBuffer bb = ByteBuffer.allocate(Integer.BYTES * 3);
+    try (RetriableS3OutputStream out = new RetriableS3OutputStream(
+        config,
+        s3,
+        path,
+        false
+    )) {
+      bb.clear();
+      bb.putInt(1);
+      bb.putInt(2);
+      bb.putInt(3);
+      out.write(bb.array());
+    }
+    // each chunk 10 bytes, so there should be 2 chunks.
+    Assert.assertEquals(2, s3.partRequests.size());
+    s3.assertCompleted(chunkSize, Integer.BYTES * 3);
+  }
+
+  @Test
+  public void testWriteSmallBufferShouldSucceed() throws IOException
+  {
+    maxResultsSize = 1000;
+    chunkSize = 128;
+    try (RetriableS3OutputStream out = new RetriableS3OutputStream(
+        config,
+        s3,
+        path,
+        false
+    )) {
+      for (int i = 0; i < 600; i++) {
+        out.write(i);
+      }
+    }
+    // each chunk 128 bytes, so there should be 5 chunks.
+    Assert.assertEquals(5, s3.partRequests.size());
+    s3.assertCompleted(chunkSize, 600);
+  }
+
+  @Test
+  public void testHitResultsSizeLimit() throws IOException
+  {
+    maxResultsSize = 50;
+    ByteBuffer bb = ByteBuffer.allocate(Integer.BYTES);
+    try (RetriableS3OutputStream out = new RetriableS3OutputStream(
```

Comment:
Thanks for using try-with-resources.

Having the client create the S3 output stream seems to couple the implementation to the client a bit too much. This is where that "connector" idea would be better:

`OutputStream out = s3Connector.openForWrite(path)`. that way, the connector, not the client, manages the config and s3, and chooses the correct stream type. This lets us more easily improve the streams without locking the client into our implementation.

Reply:
Arent we doing that now?
As this test case specifically checks the chunking logic for the Retryable OutputStream hence not going throught the connector.